File: 3310-remove-methods-from-project/3310-remove-methods-from-project.cpp
class Solution {
public:
    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        vector<vector<int>> adj(n);

        for(auto& e : invocations) adj[e[0]].push_back(e[1]);
        vector<char> sus(n, 0);
        vector<int> stk = {k};
        sus[k] = 1;
        while (!stk.empty()) {
            int u = stk.back(); stk.pop_back();
            for (int v: adj[u]) if (!sus[v]) {sus[v] = 1; stk.push_back(v); }
        }
        for (auto& e: invocations) {
            if (!sus[e[0]] && sus[e[1]]) {
                vector<int> res(n);
                for (int i = 0 ; i < n; i++) res[i] = i;
                return res;
            }
        }
        vector<int> res;
        for(int i=0; i<n; i++) if (!sus[i]) res.push_back(i);
        return res;
    }
};

File: 3310-remove-methods-from-project/3310-remove-methods-from-project.py
class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(n)]
        for a, b in invocations:
            adj[a].append(b)

        # BFS from k
        suspicious = set()
        q = deque([k])
        suspicious.add(k)

        while q:
            node = q.popleft()
            for neighbor in adj[node]:
                if neighbor not in suspicious:
                    suspicious.add(neighbor)
                    q.append(neighbor)

        for a, b in invocations: 
            if a not in suspicious and b in suspicious:
                return list(range(n))


        return [i for i in range(n) if i not in suspicious]
        
        